### ISSUE TITLE
ArrayIndexOutOfBounds in CutCreator

### ISSUE THREAD

**Issue by jhoenicke** [ultimate-pa/smtinterpol#19]
[This file](http://www.informatik.uni-freiburg.de/~hoenicke/cutcreatorbug.smt2) (217 MB!) produces ArrayIndexOutOfBounds in CutCreator.  The fault seems to be in the incremental engine though. The tableaux is broken when the bug occurs. It contains a variable in the matrix that is no longer in the list of linear variables.

**Comment by psychon** [comment]
I let the  [delta tool](https://manpages.debian.org/unstable/delta/singledelta.1.en.html) play a bit with the 217 MB file and the following interestingness script:
```sh
#!/bin/sh
(echo '(set-option :print-success false)(set-option :verbosity 3)' ; cat "$1" ) | java -jar /tmp/smtinterpol.jar 2>&1 | grep ArrayIndexOutOfBoundsException
```
The result has just 1.4 MiB and can be found here: https://gist.github.com/psychon/df8b458836d62b56d1e650f5e5b7ba58

Example:
```
$ curl https://gist.githubusercontent.com/psychon/df8b458836d62b56d1e650f5e5b7ba58/raw/92e8a7dce6d336eef2a67d091c8d2e63127f46d8/array-exception.smt | java -jar smtinterpol.jar
[...]
FATAL - Unknown exception during check
java.lang.ArrayIndexOutOfBoundsException: -1
	at de.uni_freiburg.informatik.ultimate.smtinterpol.theory.linar.CutCreator.generateCuts(CutCreator.java:745)
	at de.uni_freiburg.informatik.ultimate.smtinterpol.theory.linar.LinArSolve.ensureIntegrals(LinArSolve.java:1226)
	at de.uni_freiburg.informatik.ultimate.smtinterpol.theory.linar.LinArSolve.computeConflictClause(LinArSolve.java:605)
	at de.uni_freiburg.informatik.ultimate.smtinterpol.dpll.DPLLEngine.checkConsistency(DPLLEngine.java:901)
	at de.uni_freiburg.informatik.ultimate.smtinterpol.dpll.DPLLEngine.solve(DPLLEngine.java:1039)
	at de.uni_freiburg.informatik.ultimate.smtinterpol.smtlib2.SMTInterpol$CheckType$1.check(SMTInterpol.java:98)
	at de.uni_freiburg.informatik.ultimate.smtinterpol.smtlib2.SMTInterpol.checkSat(SMTInterpol.java:759)
	at de.uni_freiburg.informatik.ultimate.smtinterpol.smtlib2.Parser$Action$.CUP$do_action(Parser.java:1963)
	at de.uni_freiburg.informatik.ultimate.smtinterpol.smtlib2.Parser.do_action(Parser.java:853)
	at java_cup.runtime.LRParser.parse(Unknown Source)
	at de.uni_freiburg.informatik.ultimate.smtinterpol.smtlib2.ParseEnvironment.parseStream(ParseEnvironment.java:129)
	at de.uni_freiburg.informatik.ultimate.smtinterpol.smtlib2.ParseEnvironment.parseScript(ParseEnvironment.java:114)
	at de.uni_freiburg.informatik.ultimate.smtinterpol.smtlib2.SMTLIB2Parser.run(SMTLIB2Parser.java:40)
	at de.uni_freiburg.informatik.ultimate.smtinterpol.Main.main(Main.java:182)
```

**Comment by psychon** [comment]
Actually... scratch that. I was using `smtinterpol-2.1-219-g60cc1b6` for my test. With `smtinterpol-2.1-333-g717fe65`, my minimisation result no longer reproduces the problem. However, the original 217 MB file does not do so either.

I guess this means that the bug was (accidentally) fixed in the last year?

**Comment by psychon** [comment]
A quick git bisect says that the fix is commit 2dfa21a07394c4eb5a98452b289507e6bb13f908. However, that feels to me like it just hides the bug and does not actually fix anything.

**Comment by psychon** [comment]
(Sorry for the spam)

Yup, latest master with the following patch applied and with my minified reproducer file results in:
```
[...]
success
Unexpected Exception: java.lang.ArrayIndexOutOfBoundsException: -1
(error "java.lang.ArrayIndexOutOfBoundsException: -1")
```
```diff
diff --git a/SMTInterpol/src/de/uni_freiburg/informatik/ultimate/smtinterpol/smtlib2/SMTInterpol.java b/SMTInterpol/src/de/uni_freiburg/informatik/ultimate/smtinterpol/smtlib2/SMTInterpol.java
index 99f4562..45da2c4 100644
--- a/SMTInterpol/src/de/uni_freiburg/informatik/ultimate/smtinterpol/smtlib2/SMTInterpol.java
+++ b/SMTInterpol/src/de/uni_freiburg/informatik/ultimate/smtinterpol/smtlib2/SMTInterpol.java
@@ -707,7 +707,7 @@ public class SMTInterpol extends NoopScript {
                                mBy0Seen = mStackLevel;
                        }
                        if (mNumAsserts++ >= mNextQuickCheck) {
-                               mNextQuickCheck *= 2;
+                               mNextQuickCheck = -42;
                                if (!mEngine.quickCheck()) {
                                        mLogger.info("Assertion made context inconsistent");
                                        return LBool.UNSAT;
```

**Comment by psychon** [comment]
I promise that this is my last reply-spam

Since this `delta`-tool works by removing whole lines, I took the minimised example file and added extra line breaks where it logically made sense. The result was minimised some more and is now small enough for inline posting (after removing line breaks again and fixing up some syntax errors(???))
```
(set-logic QF_AUFLIRA)
(declare-sort channelID 0)
(declare-fun chan_TX_START () channelID)
(declare-fun c_sync_channel () channelID)
(declare-fun c_queue$new () Bool)
(declare-fun c_queue$reset () Bool)
(declare-fun c_MAC_1$t () Real)
(declare-fun c_MAC_1$t_primed () Real)
(declare-fun c_MAC_1$t$reset () Bool)
(declare-fun c_MAC_1$t0_primed () Real)
(declare-fun c_MAC_1$t0$new () Real)
(declare-fun c_MAC_1$t0$reset () Bool)
(declare-fun c_MAC_1$power () Real)
(declare-fun c_MEDIUM_1$t0_primed () Real)
(declare-fun c_ENV_1$t () Real)
(declare-fun c_ENV_1$t_primed () Real)
(declare-fun v_MEDIUM_1$sendCount$new_83_const_836531876 () Int)
(declare-fun v_MEDIUM_1$sendCount$new_82_const_836531877 () Int)
(declare-fun v_MEDIUM_1$sendCount_112_const_1223638505 () Int)
(declare-fun LBE652_const_-2056445330 () Bool)
(declare-fun v_MEDIUM_1$sendCount_114_const_1223638503 () Int)
(assert (= v_MEDIUM_1$sendCount$new_82_const_836531877 v_MEDIUM_1$sendCount_112_const_1223638505))
(assert (or (not LBE652_const_-2056445330) (= v_MEDIUM_1$sendCount_114_const_1223638503 v_MEDIUM_1$sendCount$new_83_const_836531876)))
(assert (or (= v_MEDIUM_1$sendCount_112_const_1223638505 (+ v_MEDIUM_1$sendCount$new_83_const_836531876 1)) (! (and (or (= c_MAC_1$t_primed c_MAC_1$t) c_MAC_1$t$reset) (or (not c_MAC_1$t0$reset) (= c_MAC_1$t0_primed c_MAC_1$t0$new)) (or (not c_queue$reset) (not c_queue$new)) (or (= c_MEDIUM_1$t0_primed 0.0) (= c_sync_channel chan_TX_START)) (<= (* (/ 1.0 30.0) c_MAC_1$power) (+ c_ENV_1$t (/ 4145.0 768.0)))) :named precondition)))
(push 1)
(assert (and (<= (* 200.0 c_MAC_1$power) c_ENV_1$t_primed) (<= c_MAC_1$t_primed c_ENV_1$t_primed)))
(check-sat)
(pop 1)
(assert (and (<= c_MAC_1$power (+ (* 50.0 c_MEDIUM_1$t0_primed) (/ 97525.0 128.0))) (<= (+ c_MAC_1$power (/ 30475.0 128.0)) 50.0)))
(check-sat)
```

**Comment by jhoenicke** [comment]
I think I found the reason for this and for the CutCreator problem.  There is now a new version `2.5-5-gb4ee0848` online.

I decided to remove the m_simps matrix which may negatively affect LRA performance.  I think the potential bugs are more important than a bit of performence.  My experiments to measure performance are a bit unconclusive.  There are almost as many instances that run faster without the m_simps matrix, so it can be just luck (these changes have a side-effect on the decision and pivoting order).
